src/client/executor.rs

```diff
@@ -76,38 +96,69 @@ impl Client {
     /// session. Retries the operation upon failure if retryability is supported.
     async fn execute_operation_with_retry<T: Operation>(
         &self,
-        op: T,
+        mut op: T,
         mut session: Option<&mut ClientSession>,
     ) -> Result<T::O> {
-        let server = self.select_server(op.selection_criteria()).await?;
+        let server = match self.select_server(op.selection_criteria()).await {
+            Ok(server) => server,
+            Err(mut err) => {
+                err.add_labels(None, &session, None)?;
+                return Err(err);
+            }
+        };
+
+        let mut conn = match server.pool.check_out().await {
+            Ok(conn) => conn,
+            Err(mut err) => {
+                err.add_labels(None, &session, None)?;
+                return Err(err);
+            }
+        };
 
-        let mut conn = server.pool.check_out().await?;
+        // If the current transaction has been committed/aborted and it is not being
```

Comment:
This needs to be moved above server selection and connection check out to ensure the state gets updated properly and that the appropriate labels are attached.